src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -116,59 +116,96 @@ internal interface ISelfCalibratingPredictor
         IPredictor Calibrate(IChannel ch, IDataView data, ICalibratorTrainer caliTrainer, int maxRows);
     }
 
-    public abstract class CalibratedPredictorBase :
+    /// <summary>
+    /// <see cref="IWeaklyTypedCalibratedModelParameters"/> provides a weekly-typed way to access strongly-typed
+    /// <see cref="CalibratedModelParametersBase{TSubPredictor, TCalibrator}.SubModel"/> and
+    /// <see cref="CalibratedModelParametersBase{TSubPredictor, TCalibrator}.Calibrator"/>.
+    /// <see cref="IWeaklyTypedCalibratedModelParameters"/> is commonly used in weekly-typed expressions. The
+    /// existence of this interface is just for supporting existing codebase, so we discourage its uses.
+    /// </summary>
+    [BestFriend]
+    internal interface IWeaklyTypedCalibratedModelParameters
+    {
+        IPredictorProducing<float> WeeklyTypedSubModel { get; }
+        ICalibrator WeeklyTypedCalibrator { get; }
+    }
+
+    /// <summary>
+    /// Class for allowing a post-processing step, defined by <see cref="Calibrator"/>, to <see cref="SubModel"/>'s
+    /// output. For example, in binary classification, <see cref="Calibrator"/> can convert support vector machine's
+    /// output value to the probability of belonging to the positive (or negative) class. Detailed math materials
+    /// can be found at <a href="https://www.csie.ntu.edu.tw/~cjlin/papers/plattprob.pdf">this paper</a>.
+    /// </summary>
+    /// <typeparam name="TSubModel">Type being calibrated.</typeparam>
+    /// <typeparam name="TCalibrator">Type used to calibrate.</typeparam>
+    public abstract class CalibratedModelParametersBase<TSubModel, TCalibrator> :
         IDistPredictorProducing<float, float>,
         ICanSaveInIniFormat,
         ICanSaveInTextFormat,
         ICanSaveInSourceCode,
         ICanSaveSummary,
-        ICanGetSummaryInKeyValuePairs
+        ICanGetSummaryInKeyValuePairs,
+        IWeaklyTypedCalibratedModelParameters
+        where TSubModel : class, IPredictorProducing<float>
+        where TCalibrator : class, ICalibrator
     {
         protected readonly IHost Host;
 
-        public IPredictorProducing<float> SubPredictor { get; }
-        public ICalibrator Calibrator { get; }
-        public PredictionKind PredictionKind => SubPredictor.PredictionKind;
+        // Strongly-typed members.
+        /// <summary>
+        /// <see cref="SubModel"/>'s output would calibrated by <see cref="Calibrator"/>.
+        /// </summary>
+        public TSubModel SubModel { get; }
+        /// <summary>
+        /// <see cref="Calibrator"/> is used to post-process score produced by <see cref="SubModel"/>.
+        /// </summary>
+        public TCalibrator Calibrator { get; }
+
```

Comment:
Do we ever expect the user to manipulate a Calibrator directly (as in create them through their constructors)?

I noticed that the calibrators (classes implementing `ICalibrator`) are public, and their constructors are public too. But I am yet to find a place where they are instantiated directly.
Should we assume that they should only be accessible through the `CalibratorEstimators`?

Reply:
Along these lines, do we want to expose this here in the `CalibratedModelParametersBase`?

---
In reply to: [255187091](https://github.com/dotnet/machinelearning/pull/2435#discussion_r255187091) [](ancestors = 255187091)

Reply:
Any thoughts @tfinley@gmail.com ?

---
In reply to: [255187736](https://github.com/dotnet/machinelearning/pull/2435#discussion_r255187736) [](ancestors = 255187736,255187091)

Reply:
This is a worthy question, but maybe we can solve this past this issue, what you think @artidoro ?

Reply:
Sounds good!

---
In reply to: [255197677](https://github.com/dotnet/machinelearning/pull/2435#discussion_r255197677) [](ancestors = 255197677)

Reply:
If you could file an issue about it @artidoro that might be helpful, I'm not entirely certain I understand the scenario.